source/Applications/Advanced/TransformViaArucoMarker/TransformViaArucoMarker.cpp

```diff
@@ -0,0 +1,315 @@
+
+        std::vector<cv::Point2f> markerCenters2D;
+        markerCenters2D.reserve(markers.size());
+
+        const auto corner0 = markers[0];
+        const auto corner1 = markers[1];
+        const auto corner2 = markers[2];
+        const auto corner3 = markers[3];
+
+        // Fit line between two diagonal corners
+        const auto backDiagonal{ fitLine(corner2, corner0) };
+        const auto forwardDiagonal{ fitLine(corner3, corner1) };
+
+        // Find intersection of the two lines
+        // a1*x + b1 = a2*x + b2
+        // x = (b2-b1) / (a1-a2)
+        const auto xCoord =
+            (forwardDiagonal.intercept - backDiagonal.intercept) / (backDiagonal.slope - forwardDiagonal.slope);
+        const auto yCoord = backDiagonal.slope * xCoord + backDiagonal.intercept;
+        markerCenters2D.emplace_back(cv::Point2f(xCoord, yCoord));
+
+        return markerCenters2D;
+    }
+
+    std::vector<cv::Point2f> get2DMarkerCorners(const aruconano::Marker &markers)
+    {
+        if(markers.empty())
+        {
+            std::cout << "No markers detected";
+            return {};
+        }
+
+        std::vector<cv::Point2f> markerCorners2D;
+        markerCorners2D.reserve(markers.size() * 4);
+
+        markerCorners2D.emplace_back(markers[0]);
+        markerCorners2D.emplace_back(markers[1]);
+        markerCorners2D.emplace_back(markers[2]);
+        markerCorners2D.emplace_back(markers[3]);
+
+        return markerCorners2D;
+    }
+
+    std::vector<cv::Point3d> estimate3DMarkerPoints(const Zivid::PointCloud &pointCloud,
+                                                    const std::vector<cv::Point2f> &markerPoints2D)
+    {
+        if(markerPoints2D.empty())
+        {
+            return {};
+        }
+
+        const int width = pointCloud.width();
+        const int height = pointCloud.height();
+        const auto points = pointCloud.copyPointsXYZ();
+
+        std::vector<cv::Point3d> markerPoints3D;
+        markerPoints3D.reserve(markerPoints2D.size());
+        for(const auto &point2D : markerPoints2D)
+        {
+            const float x = point2D.x;
+            const float y = point2D.y;
+
+            const int x1 = clamp(static_cast<int>(std::floor(x)), 0, width - 1);
+            const int x2 = clamp(static_cast<int>(std::ceil(x)), 0, width - 1);
+            const int y1 = clamp(static_cast<int>(std::floor(y)), 0, height - 1);
+            const int y2 = clamp(static_cast<int>(std::ceil(y)), 0, height - 1);
+
+            // estimate the 3d center of the marker using bilinear interpolation
+            // See https://en.wikipedia.org/wiki/Bilinear_interpolation
+            const auto q11 = points(y1, x1);
+            const auto q12 = points(y2, x1);
+            const auto q21 = points(y1, x2);
+            const auto q22 = points(y2, x2);
+
+            const float a = (x2 - x) / (x2 - x1);
+            const float b = (x - x1) / (x2 - x1);
+
+            // linear interpolation in x direction
```

Comment:
// linear interpolation in x direction vs // linear interpolation in y-axis - make it consistent, use axis or direction?

Reply:
agree, done